Fix kubectl for namespaced users

mkulke opened the pull request:
This is an attempt at fixing #13097. Currently users locked in a namespace by the ABAC authorization module cannot use kubectl at all, since kubectl does a version negotiation using /api which is not covered by the ABAC ontologies. So I ended up creating a MetaResource type, representing "/api" and "/healthz" currently. When a user has set "kubectl: true" as a property in the authorization policy file, calls to "/api" by that user will be allowed.

Commit:
Use "nonResourcePath" property in policy to expose custom endpoints.

mkulke commented:
Signed the CLA.

ncdc commented:
@liggitt

liggitt commented on docs/admin/authorization.md:
```diff
@@ -85,6 +85,7 @@ Each line is a "policy object".  A policy object is a map with the following pro
       operations.
   - `resource`, type string; a resource from an URL, such as `pods`.
   - `namespace`, type string; a namespace string.
+  - `kubectl`, type boolean; when true, the user is allowed to access the /api endpoint (required for kubectl).
```
We shouldn't start adding special cases to these attributes. I would be ok with a `nonResourceURL` or `nonResourcePath` attribute, so policy could address non-API things.

liggitt replied:
Also, we don't want policy being oriented to the "user agent", we want it oriented to the operation being done

liggitt commented on pkg/apiserver/handlers.go:
```diff
@@ -378,6 +378,16 @@ func (r *requestAttributeGetter) GetAttribs(req *http.Request) authorizer.Attrib
 	attribs.Namespace = apiRequestInfo.Namespace
 
+	// If the raw path is just a single item, the request is to a special resource.
+	if len(apiRequestInfo.Raw) == 1 {
```
I'd rather see something like this:
```
if len(attributes.Resource) == 0 && len(attribs.Namespace) == 0 {
  attributes.nonResourcePath = req.URL.Path
}
```

liggitt commented on pkg/apiserver/handlers.go:
```diff
@@ -378,6 +378,16 @@ func (r *requestAttributeGetter) GetAttribs(req *http.Request) authorizer.Attrib
+	// If the raw path is just a single item, the request is to a special resource.
+	if len(apiRequestInfo.Raw) == 1 {
+		switch apiRequestInfo.Raw[0] {
+			case "api":
```
I don't want to special case non-resource urls here... what about `/swaggerapi`, `/`, etc? Ideally, policy would be able to address all of those things consistently

liggitt commented on pkg/auth/authorizer/interfaces.go:
```diff
@@ -56,12 +59,20 @@ func (f AuthorizerFunc) Authorize(a Attributes) error {
+	ReadOnly  		bool
+	Namespace 		string
+	Resource  		string
+	MetaResource 	MetaResource
```
@deads2k thoughts on preferred terminology for a field in authorizer.Attributes to convey a non-resource path?

deads2k replied:
In OpenShift, we settled on `NonResourcePath` (the path for the request) and `Verb`.  This allowed us to express things like, "you can do GETs on `/healthz` and `/healthz/*`".

Enumerating every available endpoint is a losing battle once we you want to subdivide things.  For example, what if we decided to subdivide health to support individual component health.  It also doesn't really buy much control since anyone wanting to write such a policy it has to know exactly they want anyway.

smarterclayton replied:
Do we have a time period to upstream this?

deads2k commented on pkg/auth/authorizer/interfaces.go:
```diff
@@ -56,12 +59,20 @@ func (f AuthorizerFunc) Authorize(a Attributes) error {
+type MetaResource int
+
+const (
+	ApiVersions MetaResource = 1 << iota
```
I would recommend against enumerating constants for policy like this.  It makes you sensitive to any refactoring order change and precludes reasonable extension by any downstream component.  String constants will avoid those problems.

mkulke commented:
Makes sense. If we want to handle this in a more generic fashion, this could be simplified further:

    {"user":"alice", "readonly": true, "customEndpoint": "/api"}
    {"user":"chuck", "customEndpoint": "/healthz"}

URL.Path is carried along in the authorizer attribs and the first thing which is evaluated, customEndpoint != "" would override any Namespace & Resource setting.

I'll fix documentation when you guys think this approach works.

liggitt commented:
I'd rather see a single rule allowing "/api" for all users, e.g.:
```
{"readonly": true, "nonResourcePath": "/api"}
```

liggitt commented:
also, I think the attributes passed to an authorizer should _either_ contain resource info (namespace, resource, resource name, etc) or a request path (for non-resource requests), not both. Otherwise, we'll end up with authorizers that make path-based authorization decisions on API resources, and break when a new version of the API changes the path structure used to address the same API resource

liggitt commented on pkg/apiserver/handlers.go:
```diff
@@ -378,6 +378,9 @@ func (r *requestAttributeGetter) GetAttribs(req *http.Request) authorizer.Attrib
 	attribs.Namespace = apiRequestInfo.Namespace
 
+	// We store the full path as well for custom endpoints
+	attribs.Path = req.URL.Path
```
I think this should only be set when Namespace/Resource are empty

liggitt commented on pkg/auth/authorizer/abac/abac.go:
```diff
@@ -51,9 +51,10 @@ type policy struct {
+	Readonly  			bool   `json:"readonly,omitempty"`
+	Resource  			string `json:"resource,omitempty"`
+	Namespace 			string `json:"namespace,omitempty"`
+	CustomEndpoint	string `json:"customEndpoint,omitempty"`
```
just call it `path` or `nonResourcePath`?

kargakis commented:
@kubernetes/kubectl

mkulke commented:
> also, I think the attributes passed to an authorizer should either contain resource info (namespace, resource, resource name, etc) or a request path (for non-resource requests), not both. Otherwise, we'll end up with authorizers that make path-based authorization decisions on API resources, and break when a new version of the API changes the path structure used to address the same API resource

ok, when we explicitly exclude the paths which fall into a namespace/resource pattern, it should be called nonResourcePath to highlight that you can't just put custom urls to resources in the policy file.

liggitt commented on docs/admin/authorization.md:
```diff
@@ -67,6 +67,8 @@ A request has 5 attributes that can be considered for authorization:
         resource is the empty string.
```
change to "the resource is the empty string, and the non resource path is populated instead"

liggitt commented on docs/admin/authorization.md:
```diff
@@ -108,12 +111,17 @@ If at least one line matches the request attributes, then the request is authori
 
+### Kubectl
+
+Kubectl uses the `/api` endpoint of api-server to negotiate client/server versions. When users, who are restricted by namespaces, should be able to use the kubectl client, `/api` has to exempted from authorization checks via a `"nonResourcePath":"/api"` entry in the policy file (see example below).
```
has to *be* exempted

liggitt commented on pkg/auth/authorizer/abac/abac.go:
```diff
@@ -51,9 +51,10 @@ type policy struct {
+	Readonly  			bool   `json:"readonly,omitempty"`
+	Resource  			string `json:"resource,omitempty"`
+	Namespace 			string `json:"namespace,omitempty"`
+	NonResourcePath	string `json:"nonResourcePath,omitempty"`
```
gofmt whitespace

liggitt commented on pkg/auth/authorizer/abac/abac.go:
```diff
@@ -98,7 +99,15 @@ func NewFromFile(path string) (policyList, error) {
-	if p.subjectMatches(a) {
+
+	switch {
+	case p.NonResourcePath != "":
```
`subjectMatches` and readonly checks still apply, move the nonResourcePath check to be a peer to the resource checking

liggitt replied:
`subjectMatches` returns true for all users if `p.user` and `p.group` are both empty

liggitt replied:
not sure what the behavior should be if you set both p.NonResourcePath and p.Namespace/p.Resource... that seems like an error... I think want a single thing allowed per policy rule

liggitt commented on pkg/auth/authorizer/interfaces.go:
```diff
@@ -58,10 +61,11 @@ func (f AuthorizerFunc) Authorize(a Attributes) error {
+	ReadOnly  			bool
+	Namespace 			string
+	Resource  			string
+	NonResourcePath	string
```
gofmt whitespace

liggitt commented on pkg/auth/authorizer/interfaces.go:
```diff
@@ -41,6 +41,9 @@ type Attributes interface {
 	GetResource() string
+
+	// In case a special endpoint like /api, /healthz has been called.
+	GetNonResourcePath() string
```
document that either this or GetNamespace()/GetResource() return results, never both

mkulke commented:
I addressed the remarks and squashed a bunch of commits together, and pushed stuff again using my corporate email address (which I used in the CLA doc). 

however that does not seem to impress the the red CLA badge, is there anything else I can do on my side?

liggitt commented on pkg/auth/authorizer/abac/abac_test.go:
```diff
@@ -41,7 +41,7 @@ func TestOneLineFileNoNewLine(t *testing.T) {
 func TestTwoLineFile(t *testing.T) {
 	_, err := newWithContents(t, `{"user":"scheduler",  "readonly": true, "resource": "pods"}
-{"user":"scheduler",  "readonly": true, "resource": "services"}
+{"user":"scheduler",  "readonly": true, "kind": "services"}
```
why was this changed? isn't `resource` correct?

mkulke replied:
yup, just a merge relict, i think. (surprisingly doesn't seem to affect the tests), but in the 1.0 branch "ns" and "kind" are used everywhere and the unit tests don't complain. 

speaking of which: as the feature is additive, is there a chance to apply this to the 1.0 branch?

liggitt commented on pkg/auth/authorizer/abac/abac_test.go:
```diff
@@ -109,24 +113,29 @@ func TestNotAuthorized(t *testing.T) {
+		// though he cannot write to it
+		{User: uChuck, RO: false, NonResourcePath: "/api", Resource: "", NS: "", ExpectAllow: false},
+		// while he can write to /custom
+		{User: uChuck, RO: false, NonResourcePath: "/custom", Resource: "", NS: "", ExpectAllow: true},
```
add a test case that makes sure any user matches the readonly /api rule (User:"randomUser", etc)

mkulke replied:
"Chuck" is such a user, no? There is no policy with chuck involved.

liggitt replied:
oh, you're right, sorry

liggitt commented on pkg/auth/authorizer/abac/abac_test.go:
```diff
@@ -109,24 +113,29 @@ func TestNotAuthorized(t *testing.T) {
+		{User: uChuck, RO: false, NonResourcePath: "/custom", Resource: "", NS: "", ExpectAllow: true},
 	}
-	for i, tc := range testCases {
+	for _, tc := range testCases {
```
index is still helpful, rather than trying to eyeball the struct and comparing to test cases... I'd keep it, even if you add details to the error

liggitt commented:
not sure about the CLA question

jlowdermilk commented:
@mkulke, reply with "I signed it" to tell the bot to recheck.

mkulke commented:
I signed it

liggitt commented on pkg/auth/authorizer/abac/abac_test.go:
```diff
@@ -56,15 +56,18 @@ func TestExampleFile(t *testing.T) {
+		 {"user":"kubelet", 										"resource": "events"}
+		 {"user":"alice", 																							"namespace": "projectCaribou"}
+		 {"user":"bob", 			"readonly": true, 												"namespace": "projectCaribou"}`)
+
```
fixup whitespace... tabs are blowing up the line length and alignment...

mkulke commented:
I signed it!

jlowdermilk commented:
@mkulke sorry, it doesn't appear to be on file. Can check that it shows up at https://cla.developers.google.com/clas?

mkulke commented:
@jlowdermilk it doesn't show up there. i guess, that's because i did not use my (private) gmail-address in the CLA doc but my corporate-email - i got that in my corporate inbox, signed it and when i check it again, it says it's signed.

the bot said: 

> If you've already signed a CLA, it's possible we don't have your GitHub username or you're using a different email address. Check your existing CLA data and verify that your email is set on your git commits.

that's why i put my corporate email-addresses into my commits.

Did i mix up things here? (we don't have corporate GH accounts at work, but join the company group with out private accounts).

jlowdermilk commented:
@mkulke, if you signed cla with a corp email, that's fine. You also need to your corp email to be associated with your github account. The [?](https://github.com/kubernetes/kubernetes/commit/5b6b9dde0a221076cd497e2b1118a3e23b656714) on commit page indicates the email is not linked with your account (update in github account settings).

mkulke commented:
I signed it!

mkulke commented:
@jlowdermilk Thx, you're right. I associated the address with my GH; I hope the bot will pick it up this time.

jlowdermilk commented:
@mkulke, sorry about the hassle. I followed up with someone on our end, and they said the corporate CLA needs to be signed by a director or higher or a lawyer before it can be approved. Individual CLA is instant approval, but if thinkstep owns the code, you'll need a corp CLA.

mikedanese commented on pkg/auth/authorizer/abac/abac.go:
```diff
@@ -5,7 +5,7 @@ Licensed under the Apache License, Version 2.0 (the "License");
 You may obtain a copy of the License at
 
-    http://www.apache.org/licenses/LICENSE-2.0
+		http://www.apache.org/licenses/LICENSE-2.0
```
undo this change

yuvipanda commented:
I treid building apiserver with this patch on and while it made the initial /api call from kubectl work, it didn't fully solve the issue.

```
yuvipanda@tools-k8s-master-01:~$ kubectl --insecure-skip-tls-verify -s https://tools-k8s-master-01.tools.eqiad.wmflabs:6443  --namespace='lolrrit-wm' --user='lolrrit-wm'  --v=10 create -f nginx.yaml
I0910 04:33:54.801931   24865 debugging.go:101] curl -k -v -XGET  -H "Authorization: Bearer osiew4AiGheel6eengeiY0od" -H "User-Agent: kubectl/v1.1.0 (linux/amd64) kubernetes/e981d74" https://tools-k8s-master-01.tools.eqiad.wmflabs:6443/api
I0910 04:33:54.841574   24865 debugging.go:120] GET https://tools-k8s-master-01.tools.eqiad.wmflabs:6443/api 200 OK in 39 milliseconds
I0910 04:33:54.841620   24865 debugging.go:126] Response Headers:
I0910 04:33:54.841632   24865 debugging.go:129]     Content-Type: application/json
I0910 04:33:54.841642   24865 debugging.go:129]     Date: Thu, 10 Sep 2015 04:33:54 GMT
I0910 04:33:54.841651   24865 debugging.go:129]     Content-Length: 32
I0910 04:33:54.841756   24865 request.go:755] Response Body: {
  "versions": [
    "v1"
  ]
}
I0910 04:33:54.847033   24865 decoder.go:144] decoding stream as YAML
I0910 04:33:54.847599   24865 debugging.go:101] curl -k -v -XGET  -H "User-Agent: kubectl/v1.1.0 (linux/amd64) kubernetes/e981d74" -H "Authorization: Bearer osiew4AiGheel6eengeiY0od" https://tools-k8s-master-01.tools.eqiad.wmflabs:6443/swaggerapi/api/v1
I0910 04:33:54.848210   24865 debugging.go:120] GET https://tools-k8s-master-01.tools.eqiad.wmflabs:6443/swaggerapi/api/v1 403 Forbidden in 0 milliseconds
I0910 04:33:54.848233   24865 debugging.go:126] Response Headers:
I0910 04:33:54.848244   24865 debugging.go:129]     Date: Thu, 10 Sep 2015 04:33:54 GMT
I0910 04:33:54.848254   24865 debugging.go:129]     Content-Length: 31
I0910 04:33:54.848262   24865 debugging.go:129]     Content-Type: text/plain; charset=utf-8
I0910 04:33:54.848283   24865 request.go:755] Response Body: Forbidden: "/swaggerapi/api/v1"
I0910 04:33:54.848364   24865 request.go:812] Response Body: Forbidden: "/swaggerapi/api/v1"
I0910 04:33:54.848729   24865 debugging.go:101] curl -k -v -XGET  -H "Authorization: Bearer osiew4AiGheel6eengeiY0od" -H "User-Agent: kubectl/v1.1.0 (linux/amd64) kubernetes/e981d74" https://tools-k8s-master-01.tools.eqiad.wmflabs:6443/swaggerapi/experimental/v1
I0910 04:33:54.849168   24865 debugging.go:120] GET https://tools-k8s-master-01.tools.eqiad.wmflabs:6443/swaggerapi/experimental/v1 403 Forbidden in 0 milliseconds
I0910 04:33:54.849200   24865 debugging.go:126] Response Headers:
I0910 04:33:54.849206   24865 debugging.go:129]     Date: Thu, 10 Sep 2015 04:33:54 GMT
I0910 04:33:54.849544   24865 debugging.go:129]     Content-Length: 40
I0910 04:33:54.849550   24865 debugging.go:129]     Content-Type: text/plain; charset=utf-8
I0910 04:33:54.849564   24865 request.go:755] Response Body: Forbidden: "/swaggerapi/experimental/v1"
I0910 04:33:54.849617   24865 request.go:812] Response Body: Forbidden: "/swaggerapi/experimental/v1"
F0910 04:33:54.849810   24865 helpers.go:71] error validating "nginx.yaml": error validating data: the server does not allow access to the requested resource; if you choose to ignore these errors, turn validation off with --validate=false
```

Neither `{"readonly": true, "nonResourcePath": "/swaggerapi/experimental/v1"}` nor `{"readonly": true, "nonResourcePath": "/swaggerapi"}`` helped

yuvipanda commented:
Works fine if I pass -validate=false

yuvipanda commented:
That's just for create - get / describe works fine regardless.

liggitt commented:
kubectl defaults to verify=true and downloads swaggerspec for every create command when verifying? that's a pretty heavyweight interaction…
will need to dig into why that rule didn't cover it. Also makes me wonder if a "/path/*" form is needed

liggitt commented:
Did you restart the API server after updating the abac rules to include swaggerapi?

mkulke commented:
@jlowdermilk: thx for the clarification, I sent the thing to my boss, he'll be able to sign it on monday.

@liggitt, @yuvipanda: hmm, I can create objects in a namespace w/o problems:

authorization.jsonl:

    {"user": "admin"}
    {"readonly": true, "nonResourcePath": "/api"}
    {"user": "scheduler", "readonly": true, "resource": "pods"}
    {"user": "scheduler", "resource": "bindings"}
    {"user": "kubelet", "resource": "pods"}
    {"user": "kubelet", "readonly": true, "resource": "services"}
    {"user": "kubelet", "readonly": true, "resource": "endpoints"}
    {"user": "kubelet", "resource": "events"}
    {"user": "kubelet", "resource": "nodes"}
    {"user": "kube-proxy", "readonly": true, "resource": "services"}
    {"user": "kube-proxy", "readonly": true, "resource": "endpoints"}
    {"user": "kube-proxy","resource": "events"}
    {"user": "red", "namespace": "red"}

kubectl_config.yaml:

    apiVersion: v1
    clusters:
    - cluster:
        server: https://k8s-mgns47.abc.io
      name: mgns
    contexts:
    - context:
        cluster: mgns
        namespace: red
      user: red
    current-context: mgns-red
    kind: Config
    preferences: {}
    users:
    - name: red
      user:
        token: ABCD1234

reds-nginx.yaml:

    apiVersion: v1
    kind: ReplicationController
    metadata:
      labels:
        run: nginx
      name: reds-nginx
    spec:
      replicas: 1
      selector:
        run: reds-nginx
      template:
        metadata:
          labels:
            run: reds-nginx
        spec:
          containers:
          - image: nginx
            name: nginx

kubectl output

    $ kubectl get po
    NAME      READY     STATUS    RESTARTS   AGE
    
    $ kubectl create -f reds-nginx.yaml 
    replicationcontrollers/reds-nginx
    
    $ kubectl get rc
    CONTROLLER   CONTAINER(S)   IMAGE(S)   SELECTOR         REPLICAS
    reds-nginx   nginx          nginx      run=reds-nginx   1

what about this --validate=true default? in the head documentation I don't find anything about it and my current 1.0.3 kubectl claims that --validate=false is the default.

liggitt commented on pkg/auth/authorizer/interfaces.go:
```diff
@@ -41,6 +41,10 @@ type Attributes interface {
+
+	// In case a special endpoint like /api, /healthz has been called. This can
+	// only yield a value when both GetNamespace() & GetResource() return none.
+	GetNonResourcePath() string
```
@deads2k thoughts on the name? My desire for a nice name makes we want `GetPath()` (with documentation that it is only set for non-API requests), but my desire for complete clarity makes me want `GetNonResourcePath()`

liggitt commented:
Changed to true in https://github.com/kubernetes/kubernetes/issues/13806
I opened https://github.com/kubernetes/kubernetes/issues/13806 because this adds a sizeable overhead to invocations

yuvipanda commented:
@liggitt Nope, even after restarting, it doesn't work without `--validate=false`. I'm on the v1.1.0-alpha.1 tag.

mkulke commented:
> @liggitt Nope, even after restarting, it doesn't work without --validate=false. I'm on the v1.1.0-alpha.1 tag.

Seeing the same, the abac unit tests are fine with `/swaggerapi/experimental/v1`, probably the heuristics in `apiserver/handlers.go` don't work.

mkulke commented:
@liggitt ,@yuvipanda:

actually, i don't think it's broken:

The swagger paths are just mapped by another handler, so calls to `swaggerapi/experimental/v1` will legitimately become:

    DEBUG: apiRequestInfo: {Verb:get APIVersion: Namespace: Resource:swaggerapi Subresource:v1 Kind: Name:experimental Parts:[swaggerapi experimental v1] Raw:[]}`

Thus adding `{"readonly": true, "resource": "swaggerapi"}` does the job.

    # Token ABCD1234 resolves to a namespaced user "red" (see comments above).
    $ http https://k8s-mgns49.thinkstep.io/swaggerapi/experimental/v1 "Authorization: Bearer ABCD1234"
    HTTP/1.1 404 Not Found
    Content-Length: 24
    Content-Type: text/plain; charset=utf-8
    Date: Thu, 10 Sep 2015 18:25:26 GMT

    ApiDeclaration not found

Now i can create objects as a namespaced user:

    $ ./kubectl-dev create --validate=true -f bla.yaml 
    replicationcontroller "reds-nginx" created

Should I add the `resource: "swaggerapi"` thing to the kubectl section in the doc? Or will that requirement change? (It does indeed retrieve the whole swagger json in create operations)

liggitt commented:
pretty sure that's actually a bug in the `GetAPIRequestInfo` function... it thinks it has enough path segments to identify a resource, but it's just assigning path segments arbitrary meanings

liggitt commented:
`GetAttribs` should probably only call `apiRequestInfoResolver.GetAPIRequestInfo` if the request falls under one of the `apiRoots` provided to `NewRequestAttributeGetter`

mkulke commented:
@liggitt: yup, you're right. i thought there might be a custom resolver for swagger, but it just made sense by accident, i guess. So now the path is only evaluated when the first path fragment is part of `APIPrefixes`. Can we safely assume that an api prefix will be just a single path fragment?

liggitt commented:
No, new API work in progress makes that assumption dangerous

liggitt commented on pkg/apiserver/handlers.go:
```diff
@@ -367,16 +367,35 @@ func (r *requestAttributeGetter) GetAttribs(req *http.Request) authorizer.Attrib
+			break
+		}
+		i++
+	}
```
I get nervous when I see code like this without accompanying tests. Should also test trailing slash handling (allowing `/api` vs `/api/`)

mkulke commented:
@liggitt sure, added some tests. Can you peek over those to make sure they're doing the right thing? (not so sure about whether `/api/`should count as prefixed path)

smarterclayton commented:
@liggitt is out this week, @deads2k if you have a few minutes to look at tests I would appreciate it.

deads2k commented on docs/admin/authorization.md:
```diff
@@ -64,9 +64,11 @@ A request has 5 attributes that can be considered for authorization:
   - what resource is being accessed.
```
Above, "A request has **6** attributes"

deads2k commented on docs/admin/authorization.md:
```diff
@@ -108,12 +111,19 @@ If at least one line matches the request attributes, then the request is authori
+
+Kubectl uses the `/api` endpoint of api-server to negotiate client/server versions. To validate objects send to the API by create/update operations, kubectl queries certain swagger resources. For API version `v1` those would be `/swaggerapi/api/v1` & `/swaggerapi/experimental/v1`.
+
+ When users, who are restricted by namespaces, should be able to use the kubectl client, `/api` has to be exempted from authorization checks via a `"nonResourcePath":"/api"` entry in the policy file. When those users are suppossed to create or update resources, additional policies for the swagger paths `"nonResourcePath":"/swaggerapi/api/v1"`, `"nonResourcePath":"/swaggerapi/experimental/v1"` (as ...
```
That's not really what this is doing.  This isn't exempting anything from a check.  This change is granting additional permissions to users.  I'd expect something more like "You must grant readonly access to /api for all users who need to use the kubectl client (or any other negotiating client)."

deads2k replied:
Why are "users who are restricted by namespaces" a special case?  Given a `nonResourcePath`, I'd expect that every user regardless of resource request access would hit this case.

mkulke replied:
@deads2k: currently this only applies to users restricted to namespaces, AFAIK this is intended behaviour:

    func (p policy) matches(a authorizer.Attributes) bool {
      if p.subjectMatches(a) {
        if p.Readonly == false || (p.Readonly == a.IsReadOnly()) {
          switch {
          case p.NonResourcePath != "":
            if p.NonResourcePath == a.GetNonResourcePath() {
              return true
            }
          case p.Resource == "" || (p.Resource == a.GetResource()):
            if p.Namespace == "" || (p.Namespace == a.GetNamespace()) {
              return true
            }
          }
        }
      }

      return false
    }

Given a policy `{"user": "admin"}` When the user is `admin`, the policy has no `resource` restriction and no `namespace` restriction either, the action is authorized. Should we change this?

deads2k commented on pkg/apiserver/handlers.go:
```diff
@@ -354,6 +355,18 @@ func NewRequestAttributeGetter(requestContextMapper api.RequestContextMapper, re
+	has := false
+	// Slice the first / of the path off (in apiRoots they're stored w/o leading /)
+	parts := strings.Split(req.URL.Path, "/")[1:]
+	// Paths with only one fragment won't have a prefix.
```
This method name doesn't seem quite right.  It seems like the intent of this method is more `IsAPIResourceRequest`.  Otherwise, you could certainly have a request that is nothing but an APIPrefix.  This method returns false in that case.

deads2k commented on pkg/apiserver/handlers.go:
```diff
@@ -354,6 +355,18 @@ func NewRequestAttributeGetter(requestContextMapper api.RequestContextMapper, re
+	// Paths with only one fragment won't have a prefix.
+	for i := 1; i < len(parts) && !has; i++ {
+		prefix := strings.Join(parts[:i], "/")
+		has = r.apiRequestInfoResolver.APIPrefixes.Has(prefix)
```
If this is true, just return here.  It'll make the method easier to read, since you can `return false` after the loop.  (one less variable on the mental stack.)

deads2k commented on pkg/apiserver/handlers.go:
```diff
@@ -367,16 +380,24 @@ func (r *requestAttributeGetter) GetAttribs(req *http.Request) authorizer.Attrib
+	}
+
+	// If a request does not fall into an api namespace/resource pattern, it's a special path.
+	if attribs.Namespace == "" && attribs.Resource == "" {
```
I disagree with this `if` statement.  I'd rather have an `if/else` block above, so that its clear that you're either considered an `APIResourceRequest` or a `NonResourceRequest`, but never both.  I'd rather not assume that a request with the correct `APIPrefix` that fails a parse is a `NonResourcePath`.  That feels like a bad path to go down.

I could probably see an argument that says to be an `APIResourceRequest`, you must have an `APIPrefix` and an `APIGroup/APIVersion`, but I don't think keying off of `Namespace` and `Resource` is correct.

mkulke replied:
@deads2k If we can assume that `/api` is a `NonResourceRequest`, exclusive should be fine, I wasn't so sure about that.

deads2k replied:
> @deads2k If we can assume that /api is a NonResourceRequest, exclusive should be fine, I wasn't so sure about that.

I would assert that `/api` is a `NonResourceRequest` because it does not identify a resource.

deads2k commented on pkg/apiserver/handlers_test.go:
```diff
@@ -200,6 +200,35 @@ func TestTimeout(t *testing.T) {
+		// The prefix might be two or more fragments
+		{[]string{"test/123"}, "/test/123/abc", true},
+		{[]string{"test/123"}, "/test/123/", true},
+		{[]string{"test/123"}, "/test/123", false},
```
Add a test where they don't match.  Something like `{[]string{"api"}, "/version", false},`

deads2k commented on pkg/apiserver/handlers_test.go:
```diff
@@ -200,6 +200,35 @@ func TestTimeout(t *testing.T) {
 	}
 }
 
+func TestHasApiPrefix(t *testing.T) {
```
These are necessary, but not sufficient.  Enough logic changed in `GetAttribs`, that it requires tests as well.  I'm particular concerned about the current factorization that bases a resource parse on one condition and a `NonResourcePath` on another.

Making that code mutually exclusive will help inspectability, but it will still need tests.

mkulke replied:
Ok. there is no test for `GetAttribs()` yet, any hints on how to mock `requestAttributeGetter`. Are there factories for this or should I implement interfaces. A pointer to where to peek in the source code would be fine.

Also, is it considered good practice to put helper fn's on structs to group stuff? `hasApiPrefix()` does not modify the state of the `requestAttributeGetter` obj, so it could just be a func.

deads2k replied:
> hasApiPrefix() does not modify the state of the requestAttributeGetter obj, so it could just be a func.

That's a good point, it probably doesn't belong on the interface and should just be a plain method.

deads2k replied:
> Ok. there is no test for GetAttribs() yet, any hints on how to mock requestAttributeGetter. Are there factories for this or should I implement interfaces. A pointer to where to peek in the source code would be fine.

The code looks pretty tolerant of a stubbed `api.RequestContextMapper`.  I would stub the `APIRequestInfoResolver` as its done in `handlers_test.go`.

mkulke commented:
@deads2k: about those failing tests: I used a fresh clone to build & test the branch and don't see any errors. Are the ci/shippable tests rebasing automatically or something?

deads2k commented on pkg/auth/authorizer/abac/abac.go:
```diff
@@ -100,13 +101,19 @@ func NewFromFile(path string) (policyList, error) {
+				if p.NonResourcePath == a.GetNonResourcePath() {
+					return true
+				}
+			case p.Resource == "" || (p.Resource == a.GetResource()):
```
If `len(a.GetNonResourcePath()) != 0`, then this branch of the switch should never return true.

deads2k replied:
> If len(a.GetNonResourcePath()) != 0, then this branch of the switch should never return true.

And add a test to make sure that's the case.

mkulke commented:
@deads2k: Thanks a lot for your comments. I changed the behaviour, so that blank check user-policies like `{"user":"admin"}` won't automatically authorize non-resource paths anymore. However this introduces a breaking change. `{"nonResourcePath": "/api"}`, etc. policies will now be mandatory when using ABAC authz.

I think was this way because the PR started as a bolt-on backwards compatible attempt to make kubectl work with namespace restrictions. Users who currently are using ABAC and no namespaces, will have to fix their authorization files. Thus, backporting a fix to 1.0 will be hard.

deads2k commented:
> @deads2k: Thanks a lot for your comments. I changed the behaviour, so that blank check user-policies like {"user":"admin"} won't automatically authorize non-resource paths anymore. However this introduces a breaking change. {"nonResourcePath": "/api"}, etc. policies will now be mandatory when using ABAC authz.
> 
> I think was this way because the PR started as a bolt-on backwards compatible attempt to make kubectl work with namespace restrictions. Users who currently are using ABAC and no namespaces, will have to fix their authorization files. Thus, backporting a fix to 1.0 will be hard

I'm happy with the code, but the question of default permissive or default restrictive is one for @erictune .  The code in question is here: https://github.com/kubernetes/kubernetes/pull/13667/files#diff-deaf08ba14d40a64444f67c3c9f0237cR110 .  I find the either/or to match my expectations of what the rules would do.  Without the either/or, a resource based rule could authorize a nonResource request.

If it matters, in openshift we're default restrictive and we use "*" to indicate all nonResourcePaths.  That would still be a breaking change here.

liggitt commented:
There needs to be a simple way to give someone "all access", including non-resource paths

Currently, the way to grant broad API access in the ABAC file is to omit namespace/resource restrictions for a given rule. I think that's a dangerous approach, since a typo in a field name makes the JSON deserializer drop the field, effectively broadening access. I'd rather make you specify something like `"namespace":"*","resource":"*","nonResourcePath":"*"` to get all access.

@erictune thoughts on expressing `*` permissions in the ABAC rules?

liggitt commented on pkg/auth/authorizer/abac/abac.go:
```diff
@@ -100,13 +101,20 @@ func NewFromFile(path string) (policyList, error) {
+					return true
+				}
+			// When the path is a non-resource path it cannot match.
+			case len(a.GetNonResourcePath()) == 0 && (p.Resource == "" || (p.Resource == a.GetResource())):
```
I think this check should be:
```
case len(a.GetResource()) > 0 && (p.Resource == "" || (p.Resource == a.GetResource())):
```

liggitt replied:
(eventually, I'd like the check to be `p.Resource == "*"` rather than `p.Resource == ""`, but that's a breaking change that requires more discussion)

mkulke replied:
@liggitt: `case len(a.GetResource()) > 0 && ...` would be a breaking change, as well. At least it breaks the tests: `{"user":"alice","namespace":"projectCaribou"}` should be allowed to a ns w/o a resource (`{User: uAlice, RO: true, Resource: "", NS: "projectCaribou", ExpectAllow: true},`)

liggitt commented on pkg/auth/authorizer/abac/abac.go:
```diff
@@ -100,13 +101,20 @@ func NewFromFile(path string) (policyList, error) {
 		if p.Readonly == false || (p.Readonly == a.IsReadOnly()) {
-			if p.Resource == "" || (p.Resource == a.GetResource()) {
+			switch {
+			case p.NonResourcePath != "":
```
if the incoming attributes indicate a non-resource, we always want to choose this branch, right? Shouldn't the check be this:
```
case len(a.GetNonResourcePath()) > 0:
```

mkulke replied:
Actually, we circle through a set of policies and attempt to match those against the request attributes here, so i'd argue it's correct to verify whether the policy is a non-resource policy first (in policies non-resource attributes invalidate ns/resource attributes) and then check if the request is matching, no?

liggitt replied:
I think you're right

liggitt replied:
I'd like to preserve the information from `GetRequestInfo` in the attributes. If we did, this structure could look something like this:
```
// Ensure resource rules only match resource requests, and non-resource rules only match non-resource requests
isResourceRule := p.NonResourcePath == ""
if isResourceRule != a.IsResourceRequest() {
  return false
}

if isResourceRule {
  // Require namespace to match, if present in the policy rule
  if p.Namespace != "" && p.Namespace != a.GetNamespace() {
    return false
  }
  // Require resource to match, if present in the policy rule
  if p.Resource != "" && p.Resource != a.GetResource() {
    return false
  }
  return true
}

// Exact non-resource matches
if p.NonResourcePath == a.GetPath() {
  return true
}
// Subpath matches for `.../*` non-resource rules
if strings.HasSuffix(p.NonResourcePath, "/*") && strings.HasPrefix(a.GetPath(), string.TrimRight(p.NonResourcePath, "*")) {
  return true
}
return false
```

bgrant0607 commented:
Any hope for this converging in the next day or two?

mkulke commented:
@bgrant0607: I can work on the rebase tomorrow (CET).

liggitt commented on pkg/apiserver/handlers.go:
```diff
@@ -363,18 +376,24 @@ func (r *requestAttributeGetter) GetAttribs(req *http.Request) authorizer.Attrib
 	}
 
 	apiRequestInfo, _ := r.apiRequestInfoResolver.GetAPIRequestInfo(req)
```
in master, `GetAPIRequestInfo` changed to `GetRequestInfo`, and now includes these attributes:
```
	// IsResourceRequest indicates whether or not the request is for an API resource or subresource
	IsResourceRequest bool
	// Path is the URL path of the request
	Path string
	// Verb is the kube verb associated with the request for API requests, not the http verb.  This includes things like list and watch.
	// for non-resource requests, this is the lowercase http verb
	Verb string
```

liggitt replied:
I'd recommend copying those three attributes verbatim into `authorizer.Attributes`

liggitt commented on pkg/apiserver/handlers.go:
```diff
@@ -363,18 +376,24 @@ func (r *requestAttributeGetter) GetAttribs(req *http.Request) authorizer.Attrib
-	// we can extract the resource.  Otherwise, not.
-	attribs.Resource = apiRequestInfo.Resource
+	// Check whether meaningful api information can be resolved for the current path
+	if isAPIResourceRequest(r.apiRequestInfoResolver.APIPrefixes, req) {
```
this can just become `requestInfo.IsResourceRequest`

liggitt commented on pkg/apiserver/handlers_test.go:
```diff
@@ -201,6 +202,73 @@ func TestTimeout(t *testing.T) {
 	}
 }
 
+func TestIsAPIResourceRequest(t *testing.T) {
```
can drop this, covered by tests for `GetRequestInfo`

liggitt commented:
rebased and added tests and support for wildcard non-resource path suffixes in https://github.com/kubernetes/kubernetes/pull/16148

yuvipanda commented:
This no longer cleanly applies to 1.1.0-beta. So I guess this won't make 1.1?

bgrant0607 commented:
Removed from 1.1.

yuvipanda commented:
I did a super naive rebase on https://github.com/yuvipanda/kubernetes/tree/fix_kubectl_for_namespaced_users

Tagging @erictune and @pmorie too. I'm carrying an older version of this patch on my current install

mkulke commented:
@yuvipanda: thx, i really like this to land in upstream. we're relying on namespace'd users and have to run patched versions in our clusters currently as well.

yuvipanda commented:
Yes, I'm carrying patches too and since this doesn't rebase cleanly on
master I can't get on master or any newer patch. Can you rebase this less
blindly than me? :D I'll try to push to get this merged too.

mkulke commented:
@yuvipanda: sorry I forgot, @liggitt opened a PR with the necessary rebase work some weeks ago (see comments above) #16148

liggitt commented:
https://github.com/kubernetes/kubernetes/pull/16148 Is close, I am closing out our 1.1 release and will hopefully get back to it shortly

yuvipanda commented:
Awesome. In that case can you close this PR?

On Wed, Nov 11, 2015 at 2:14 PM, Jordan Liggitt <notifications@github.com>
wrote:

> #16148 <https://github.com/kubernetes/kubernetes/pull/16148> Is close, I
> am closing out our 1.1 release and will hopefully get back to it shortly
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13667#issuecomment-155926378>
> .
>



-- 
Yuvi Panda T
http://yuvi.in/blog

mkulke commented:
@yuvipanda: yup.

[closed: duplicate, work continues in #16148]
